Update image-builder documentation (#11036)
kyma-project/test-infra: cmd/image-builder/image-builder.md

# Image Builder

Image Builder is a tool for building OCI-compliant images.

> It builds images using Azure DevOps (ADO) pipeline backend.

Reviewer: ```suggestion
It builds images using the Azure DevOps (ADO) pipeline backend.
```

> It can run in two modes. The first mode is the default mode, where Image Builder acts as a client and triggers the ADO pipeline.

Reviewer 2: ```suggestion
It can run in two modes. One is the default mode, where Image Builder acts as a client and triggers the ADO pipeline.
```

> In this mode Image Builder support running as part of a GitHub Actions workflow.

Reviewer: ```suggestion
In this mode, Image Builder supports running as part of a GitHub Actions workflow.
```

> In second mode Image Builder runs as part of the `oci-image-builder` pipeline in the ADO backend.

Reviewer: ```suggestion
In the second mode, Image Builder runs as part of the `oci-image-builder` pipeline in the ADO backend.
```

> In the second mode, Image Builder runs as part of the `oci-image-builder` pipeline in the ADO backend.

Reviewer 2: ```suggestion
In the other mode, Image Builder runs as part of the `oci-image-builder` pipeline in the ADO backend.
```

> The Image Builder is build and pushed as a container image to the Google artifact registry repository.

Reviewer: ```suggestion
The Image Builder is built and pushed as a container image to the Google Artifact Registry repository.
```

> The Image Builder is built and pushed as a container image to the Google Artifact Registry repository.

Reviewer 2: ```suggestion
Image Builder is built and pushed as a container image to the Google Artifact Registry repository.
```

Key features:

* Automatically provides a default tag, which is computed based on a template provided in `config.yaml`
* Supports adding multiple tags to the image

> * Supports pushing images to google artifact registries.

Reviewer: ```suggestion
* Supports pushing images to Google Artifact Registries.
```

* Supports running in a GitHub Actions workflow.
* Supports building images using the ADO pipeline backend.

## Configuration

Image Builder is configured using a global configuration YAML file, set of environment variables, and command line flags.

### Configuration YAML File

`image-builder` requires a configuration YAML file. The file holds the global configuration for the tool and is maintained by the authors.
Use the `--config` flag to provide a path to the config YAML file.

For more information about available properties in the configuration file, refer to the [config.go](config.go) file.

Here's an example file:

```yaml
registry: eu.gcr.io/kyma-project
reproducible: true
log-format: json
cache:
  enabled: true
  cache-repo: eu.gcr.io/sap-kyma-neighbors-dev/cache
  cache-run-layers: true
```

### Environment Variables

Environment variables are mainly used to provide runtime values and configuration set by the CI/CD system.
They provide details about the context in which the tool is running.

> Here is the list of environment variables used by Image Builder:

Reviewer 2: This sentence is redundant.

Author: removed

> The Image Builder uses several environment variables, which can be grouped by their use cases as follows:

Reviewer 2: ```suggestion
Image Builder uses several environment variables, which can be grouped by their use cases as follows:
```

> - `ADO_PAT`: Personal Access Token used to authenticate with the ADO API.
> - `REPO_OWNER`: Used to extract the repository owner for the ADO pipeline.
> - `REPO_NAME`: Used to extract the repository name for the ADO pipeline.
> - `JOB_TYPE`: Determines the type of the job (presubmit or postsubmit).
> - `PULL_NUMBER`: Used when the job type is a presubmit job.
> - `PULL_BASE_SHA`: Used to fetch the base commit SHA for the image tag.
> - `PULL_PULL_SHA`: Used when the job type is a pull request.
> - `CI`: Determines if the Image Builder is running inside a CI system. If set to "true", the CI system is determined and the git state is
>   loaded accordingly.
> - `GITHUB_REPOSITORY`: Used to get the repository name when the CI system is GitHub Actions.
> - `GITHUB_EVENT_NAME`: Used to determine the job type when the CI system is GitHub Actions.
> - `GITHUB_EVENT_PATH`: Used to get the path to the event JSON file when the CI system is GitHub Actions.
> - `GITHUB_SHA`: Used to get the commit SHA when the CI system is GitHub Actions.
> - `GITHUB_REF`: Used to get the pull request number when the CI system is GitHub Actions.

Reviewer 2: ```suggestion
- **ADO_PAT**: Personal Access Token used to authenticate with the ADO API.
- **REPO_OWNER**: Used to extract the repository owner for the ADO pipeline.
- **REPO_NAME**: Used to extract the repository name for the ADO pipeline.
- **JOB_TYPE**: Determines the type of the job (presubmit or postsubmit).
- **PULL_NUMBER**: Used when the job type is a presubmit job.
- **PULL_BASE_SHA**: Used to fetch the base commit SHA for the image tag.
- **PULL_PULL_SHA**: Used when the job type is a pull request.
- **CI**: Determines if Image Builder is running inside a CI system. If set to "true", the CI system is determined and the git state is
  loaded accordingly.
- **GITHUB_REPOSITORY**: Used to get the repository name when the CI system is GitHub Actions.
- **GITHUB_EVENT_NAME**: Used to determine the job type when the CI system is GitHub Actions.
- **GITHUB_EVENT_PATH**: Used to get the path to the event JSON file when the CI system is GitHub Actions.
- **GITHUB_SHA**: Used to get the commit SHA when the CI system is GitHub Actions.
- `GITHUB_REF**: Used to get the pull request number when the CI system is GitHub Actions.
```

Reviewer 2: Please use [bold font for variables](https://github.com/kyma-project/community/blob/main/docs/guidelines/content-guidelines/03-formatting.md#code-font-and-bold-font).

> Please note that the actual usage of these environment variables may vary depending on the specific configuration and usage of the Image
> Builder.

Reviewer 2: ```suggestion
> [!NOTE]
 > The actual usage of these environment variables may vary depending on the specific configuration and usage of Image Builder.
```

### Command Line Flags

Command line flags are the main way for developers to configure the tool and provide needed values for the build process.
Check the list and description of the available flags in
the [main.go](https://github.com/kyma-project/test-infra/blob/df945b96654d60f82b9738cd98129191c5e753c8/cmd/image-builder/main.go#L668) file.

> ## Azure DevOps Build Backend (ADO)

Reviewer 2: ```suggestion
## Azure DevOps (ADO) Build Backend
```

> The Image Builder by default is used to call ADO API and trigger the `oci-image-builder` pipeline ADO pipeline.

Reviewer 2: ```suggestion
Image Builder is, by default, used to call ADO API and trigger the `oci-image-builder` ADO pipeline.
```

When using the ADO backend, Image Builder is used as a client collecting values from flags and environment variables and calling ADO API.
Image Builder triggers the `oci-image-builder` pipeline. This pipeline is responsible for processing parameters provided in a call and
building, pushing, and signing an image.

> The Image Builder passes the tag definitions and values provided by user as base64-encoded pipeline `Tags` parameter.

Reviewer: ```suggestion
The Image Builder passes the tag definitions and values provided by the user as a base64-encoded pipeline `Tags` parameter.
```

> The Image Builder passes the tag definitions and values provided by the user as a base64-encoded pipeline `Tags` parameter.

Reviewer 2: ```suggestion
Image Builder passes the tag definitions and values provided by the user as a base64-encoded pipeline **Tags** parameter.
```

Encoding the value allows for passing special characters in the tag values without the need to escape them.

> The Image Builder is used as part of the `oci-image-builder` pipeline in the ADO backend too.

Reviewer 2: ```suggestion
Image Builder is used as part of the `oci-image-builder` pipeline in the ADO backend, too.
```

It's used to execute steps responsible for generating image tags and signing images using the signify service.

Apart from building images using ADO, Image Builder also supports preview mode. In preview mode,
Image Builder does not trigger the ADO pipeline but generates a YAML file with the pipeline definition.

> Using this mode allows for the validation of the pipeline definition syntax before triggering it. To use preview mode, add

Reviewer: ```suggestion
Using this mode allows for validating the pipeline definition syntax before triggering it. To use preview mode, add
```

> Apart from building images using ADO, Image Builder also supports preview mode. In preview mode,
> Image Builder does not trigger the ADO pipeline but generates a YAML file with the pipeline definition.
> Using this mode allows for the validation of the pipeline definition syntax before triggering it. To use preview mode, add
> the `--ado-preview-run=true` flag.

Reviewer 2: ```suggestion
Apart from building images using ADO, Image Builder also supports the preview mode. In the preview mode,
Image Builder does not trigger the ADO pipeline but generates a YAML file with the pipeline definition.
Using this mode allows for the validation of the pipeline definition syntax before triggering it. To use the preview mode, add
the `--ado-preview-run=true` flag.
```

You can enable every supported signing service on repository and global levels.

See the following example of sign services configuration in the `config.yaml` file:

```yaml
sign-config:
  enabled-signers:
    '*':
      - default-signify
    org/repo:
      - repo-token-notary
  signers:
    - name: default-signify
      type: notary
      config:
        endpoint: https://notary/sign
        timeout: 5m
        retry-timeout: 10s
        secret:
          path: /path/to/secret/file/signify.yaml
          type: signify
    - name: repo-token-notary
      type: notary
      config:
        endpoint: https://repo-notary/sign
        timeout: 5m
        retry-timeout: 10s
        secret:
          path: /path/to/secret/file/token
          type: token
```

All enabled signers under `'*'` are used globally. Additionally, if a repository contains another signer configuration
in the **org/repo** key, Image Builder also uses this service to sign the image.
If the job is running in CI (Prow), it picks up the current **org/repo** value from the default Prow variables. If binary

> is running outside of CI, the `--repo` flag must be used. Otherwise, the configuration will not be used.

Reviewer 2: ```suggestion
is running outside of CI, the `--repo` flag must be used. Otherwise, the configuration is not used.
```
